pymmcore_plus/mda/_engine.py

```diff
@@ -111,34 +150,69 @@ def check_canceled():
                         time.sleep(0.5)
                     else:
                         time.sleep(to_go)
-                    to_go = go_at - (time.perf_counter() - t0)
+                    to_go = go_at - (time.perf_counter() - self._t0)
+
+        # check canceled again in case it was canceled
+        # during the waiting loop
+        if self._check_canceled():
+            return True
+        return False
+
+    def _prep_hardware(self, event: MDAEvent, waitForSystem: bool = True):
+        """
+        Set the system hardware (XY, Z, channel, exposure) as defined in the event.
+
+        Parameters
+        ----------
+        event : MDAEvent
+            The event to use for the Hardware config
+        waitForSystem : bool
+            Whether to call `core.waitForSystem()`
+        """
+        if not self._running:
+            return
+        if event.x_pos is not None or event.y_pos is not None:
+            x = event.x_pos or self._mmc.getXPosition()
+            y = event.y_pos or self._mmc.getYPosition()
+            self._mmc.setXYPosition(x, y)
+        if event.z_pos is not None:
+            self._mmc.setZPosition(event.z_pos)
+        if event.channel is not None:
+            self._mmc.setConfig(event.channel.group, event.channel.config)
+        if event.exposure is not None:
+            self._mmc.setExposure(event.exposure)
+
+        if waitForSystem:
+            self._mmc.waitForSystem()
+
+    def _finish_run(self, sequence: MDASequence):
+        """
+        To be called at the end of an acquisition.
+
+        Parameters
+        ----------
+        sequence : MDASequence
+            The sequence that was finished.
+        """
+        logger.info("MDA Finished: {}", sequence)
+        self._running = False
+        self._events.sequenceFinished.emit(sequence)
+
+    def run(self, sequence: MDASequence) -> None:
```

Comment:
i know we didn't have one before, but since you added such nice docstrings elsewhere, here too? :)

Reply:
good point. I even went ahead and added them to all the method